### DISCUSSION TITLE
Add support for device replacement

### DISCUSSION THREAD

**Discussion by thecode** [home-assistant/architecture#1088] [General]
# Support for replacing an existing device

With home assistant getting a large install base and devices that may be installed for a long period of time which may eventually malfunction the need to replace a device without loosing configuration, automations, history is very important.
A simple example is a switch that is broken, user install a new switch in the same location and would prefer that nothing will be changed. 
Currently this case require deleting the existing device and adding a new device which cause loss of automations and history.

This feature was requested multiple times:
https://community.home-assistant.io/t/wth-there-is-no-easy-and-automated-way-to-replace-devices/468863
https://community.home-assistant.io/t/add-a-method-of-replacing-devices/249462

# Implementation

The background for how to implement it came to my mind from the recent support for reconfigure flow.
Implementation is almost identical to https://github.com/home-assistant/core/pull/108794 with mainly changing the related names from "Reconfigure" to "Replacement".

Unlike reconfigure which makes sure you are reconfiguring the same device, replacement should make sure you are updating the device parameters to point to a non-existing device or suggest a new detected device as part of the flow.

Frontend will show a new option "Replace device" which will trigger `async_step_reconfigure` and from there it is up to each integration to handle the reconfigure correctly and make sure that the device ids and entity ids are not changing.

If this idea makes sense I can show an implementation for Shelly integration.

**Comment by frenck** [comment]
> Frontend will show a new option "Replace device" which will trigger async_step_reconfigure and from there it is up to each integration to handle the reconfigure correctly and make sure that the device ids and entity ids are not changing.

I don't agree re-using reconfigure for that.

I think a concern in general is... what does replacing a device entail in the user flow. I think replacing often happens afterwards. You have a device (it broken or whatever), buy a new one, pair it... and then replace it?

The giving flow in this suggestion will not work with things like ZHA or Matter, as you cannot simply exchange the connection details. 

A new and old device also may be different. Is there a minor difference in version/type that may not be a real replacement?

Which brings me to the question: What is the reason why people want to "replace" (as in swap) a device in Home Assistant? And can wat can we do to approach those specific reasons?

**Reply by thecode** [reply]
> I don't agree re-using reconfigure for that.

I might have being miss understood, I don't intend to use re-configure, the reason for this discussion is to define a new option. I only gave an example that implantation for this in core and frontend is similar to reconfigure.

> The giving flow in this suggestion will not work with things like ZHA or Matter, as you cannot simply exchange the connection details.

I agree this may not be supported by every integration, however even for ZHA or Matter it may be possible to do it by asking the user while adding the device if this is replacing an existing device, but for this discussion it is out of the scope.

> A new and old device also may be different. Is there a minor difference in version/type that may not be a real replacement?

Would be responsibility of the integration to deny device replacement if they are not the same model.

> Which brings me to the question: What is the reason why people want to "replace" (as in swap) a device in Home Assistant? And can wat can we do to approach those specific reasons?

As pointed at the first post, a device that is broken/malfunction.

**Reply by frenck** [reply]
>  but for this discussion it is out of the scope.

I think it is pretty much in scope. Zigbee is one of the biggest use cases in the smart home. I don't think we can ignore that for this.

> As pointed at the first post, a device that is broken/malfunction.

That is not what I meant. I meant we need to research the reasons why someone wants to "replace"/swap a device in Home Assistant and not just use the newly added device instead. Depending on what those reasons are, there might be different opportunities or solutions.

So I was not looking for the reason why the physical device is being swapped (could be broken, whatever reason), it is the reasons for wanting to swap them in Home Assistant. If we know those specifics, we can also look at how to approach those.

**Reply by shbatm** [reply]
> That is not what I meant. I meant we need to research the reasons why someone wants to "replace"/swap a device in Home Assistant and not just use the newly added device instead. Depending on what those reasons are, there might be different opportunities or solutions.

For me, one of the main issues is that devices use a non-user configurable `device_id` string, so anything referring to the device (i.e. Triggers in an automation) are 'hard coded' to the specific device. Even removing and reading the same device will break things and require manually editing the references. Spook does a good job catching these (thanks), but that's not native. Compare to an automation referencing an entity_id, I can just rename the new entity back to the old entity_id.

**Reply by thecode** [reply]
> So I was not looking for the reason why the physical device is being swapped (could be broken, whatever reason), it is the reasons for wanting to swap them in Home Assistant. If we know those specifics, we can also look at how to approach those.

Explained on my first post and in the following toptics:
https://community.home-assistant.io/t/wth-there-is-no-easy-and-automated-way-to-replace-devices/468863
https://community.home-assistant.io/t/add-a-method-of-replacing-devices/249462

1. Automations - keep existing device automations. A simple example I had to face with. A double relay switch which has no physical switch so it is controlled by another switch ended up in 6 automations, in my case I decided to take out the device, repair it and put it back to avoid recreating the automations, but most users will not have the knowledge to repair a faulty hardware. As pointed by @shbatm if the automation is bound to entity id it is somehow possible to fix it, but if it is bounded to a device id users can't do anything with it. 
2. History, for example an energy meter (which is very common on many devices) that has gathered information for a long time, when it is failed the user has no option to keep this history, the new device will start collecting from scratch and the history of the old device will be removed when the device is removed from HA.

If we can support transferring automations & history from a device to another somehow this can also cover ZHA/Matter. In any case while I wanted to focus on devices that uses a config entry per device at first (this is why ZHA was out of scope in my suggestion) since the solution for them is easier. It is possible to take a different approach, let the new device be added first (to be able to support ZHA for example) than when a "Replace device" operation is started on the broken device the user will be asked to select the new device and history/automations will be migrated to the new device, but I don't see how this would be easily feasible unlike doing it before the new device is added which is simple (at least as I see it).

**Reply by thecode** [reply]
My original intention was to support here exactly what was implemented by ESPHome https://github.com/home-assistant/core/pull/142507
I thought it would be more user friendly to start the replacement process from the old device, but the solution implemented in ESPHome, using a repair issue can be copied to other integrations so I don't see the reason to keep this discussion open,.

**Reply by elupus** [reply]
Dont we miss some way of starting the repair process? When replacing an working device with another working device?
